fraud/bad_encoding.go

```diff
@@ -0,0 +1,179 @@
+	position uint8,
+	isRow bool,
+	eds *rsmt2d.ExtendedDataSquare,
+	errShares [][]byte,
+) (Proof, error) {
+	var tree ErasuredNamespacedMerkleTree
+	shares := make([]*Share, len(errShares))
+
+	for index, share := range errShares {
+		if share != nil {
+			tree = NewErasuredNamespacedMerkleTree(uint64(len(errShares) / 2))
+			data := eds.Row(uint(index))
+			if isRow {
+				data = eds.Col(uint(index))
+			}
+			if err := buildTreeFromDataRoot(&tree, data, index); err != nil {
+				continue
+			}
+
+			proof, err := tree.InclusionProof(position)
+			if err != nil {
+				return nil, err
+			}
+			/*
+				NamespaceID sholud be replaced with ParitySharesNamespaceID for
+				all shares except Q0(all erasure coded shares).
+				For EDS 4x4 Q0 is [[0;0],[0;1],[1:0],[1;1]].
+				x	x	x	x
+				x	x	x	x
+				x	x	x	x
+				x	x	x	x
+			*/
+			namespaceID := share[:consts.NamespaceSize]
+			if index >= len(errShares)/2 || int(position) >= len(errShares)/2 {
+				namespaceID = consts.ParitySharesNamespaceID
+			}
+
+			shares[index] = &Share{namespaceID, share, &proof}
+		}
+	}
+
+	return &BadEncodingProof{
+		BlockHeight: height,
+		Shares:      shares,
+		isRow:       isRow,
+		Position:    position,
+	}, nil
+}
+
+// Type returns type of fraud proof
+func (p *BadEncodingProof) Type() ProofType {
+	return BadEncoding
+}
+
+// Height returns block height
+func (p *BadEncodingProof) Height() uint64 {
+	return p.BlockHeight
+}
+
+// MarshalBinary converts BadEncodingProof to []byte
+func (p *BadEncodingProof) MarshalBinary() ([]byte, error) {
+	shares := make([]*pb.Share, 0, len(p.Shares))
+	for _, share := range p.Shares {
+		shares = append(shares, share.ShareToProto())
+	}
+
+	badEncodingFraudProof := pb.BadEnconding{
+		Height:   p.BlockHeight,
+		Shares:   shares,
+		Position: uint32(p.Position),
+		IsRow:    p.isRow,
+	}
+	return badEncodingFraudProof.Marshal()
+}
+
+func (p *BadEncodingProof) Validate(header *header.ExtendedHeader) error {
+	merkleRowRoots := header.DAH.RowsRoots
+	merkleColRoots := header.DAH.ColumnRoots
+	if int(p.Position) >= len(merkleRowRoots) || int(p.Position) >= len(merkleColRoots) {
+		return errors.New("invalid fraud proof: incorrect position of bad row/col")
+	}
+	if len(merkleRowRoots) != len(merkleColRoots) {
+		return errors.New("invalid fraud proof: invalid extended header")
+	}
+	if len(merkleRowRoots) != len(p.Shares) || len(merkleColRoots) != len(p.Shares) {
+		return errors.New("invalid fraud proof: invalid shares")
+	}
+
+	roots := merkleRowRoots
+	if p.isRow {
+		roots = merkleColRoots
+	}
+
+	shares := make([][]byte, len(roots))
+
+	// verify that Merkle proofs correspond to particular shares
+	for index, share := range p.Shares {
+		if share == nil {
+			continue
+		}
+		shares[index] = share.Raw
+		if ok := share.Validate(roots[index]); !ok {
+			return fmt.Errorf("invalid fraud proof: incorrect share received at position %d", index)
+		}
+	}
+
+	// rebuild a row/col
+	codec := rsmt2d.NewRSGF8Codec()
```

Comment:
If we later switch to leopard, I hope we won't forget to change all the codec instantiations here and in core.

Actually, it makes more sense to use this constructor instead: https://github.com/celestiaorg/celestia-core/blob/cb696bbd70577c08cd8340ca8cecdff54c723388/pkg/consts/consts.go#L75

Reply:
Removed codec creation at all. Now it is passing as a function argument